net: Decouple CConnman and message serialization

theuni opened the pull request:
This contains a more complete version of #9117, as well as the changes promised to @TheBlueMatt in #8708. They're combined here because the PushMessage changes require that the SendVersion is sane. This will be my last time messing with PushMessage, and it's nearing the end of the net changes needed outside of actual net code...

The changes here:
- Don't send messages after fDisconnect is set. (feeler connections are a TODO exception).
- Don't use CSerializeData for outgoing data, which incurred a memory_cleanse for each message. Messages aren't secret, and going forward, if they are they'll be encrypted.
- Add a thin, featureless serializer to turn any set of args into a byte vector.
- Add an go-between message maker so that CConnman sees only serialized messages for pushing.
- That means that the caller is now responsible for determining the node's send version. That's a good thing, it can now move to CNodeState as later step.
- CConnman attaches the header and hashes the byte vector. @jonasschnelli: This should be very easy to adapt to bip151. A new flag SERIALIZE_BIP151_ENCRYPTED or so can be added and used similar to SERIALIZE_TRANSACTION_NO_WITNESS.

theuni committed:
net: don't send any messages before handshake or after requested disconnect

Also, send reject messages earlier in SendMessages(), so that disconnections are
processed earlier.

These changes combined should ensure that no message is ever sent after
fDisconnect is set.

theuni committed:
net: No need to check individually for disconnection anymore

theuni committed:
net: add CVectorWriter and CNetMsgMaker

CVectorWriter is useful for overwriting or appending an existing byte vector.

CNetMsgMaker is a shortcut for creating messages on-the-fly which are suitable
for pushing to CConnman.

theuni committed:
net: push only raw data into CConnman

This fixes one of the last major layer violations in the networking stack.

The network side is no longer in charge of message serialization, so it is now
decoupled from Bitcoin structures. Only the header is serialized and attached
to the payload.

rebroad commented on src/main.cpp:
```diff
@@ -6502,9 +6509,13 @@ bool SendMessages(CNode* pto, CConnman& connman)
     {
         // Don't send anything until we get its version message
-        if (pto->nVersion == 0)
+        if (!pto->fSuccessfullyConnected || pto->fDisconnect)
```
thank you!!

theuni replied:
@dcousens the version message sets fSuccessfullyConnected, but I guess that's not as clear after this change.

How about "don't send anything until we're successfully connected and ok with the peer's version" ?

dcousens replied:
I think if we are still referencing version here to explain why,  then `fSuccessfullyConnected` isn't appropriate.

theuni commented:
Grr, will fix up the Windows build problem in the morning.

venzen reviewed (commented):
netmessagemaker and handling of initial Version message in main.cpp are good improvements.

TheBlueMatt commented:
Concept ACK, though I'm not convinced there arent too many copies when making the message and passing it into CConnman, but will audit when you fix up build, etc.

theuni commented:
@TheBlueMatt there's only 1 copy (other than the serialization itself), and it comes from combining the header and payload. See the note here: https://github.com/bitcoin/bitcoin/pull/9128/files#diff-9a82240fe7dfe86564178691cc57f2f1R2590

Either way, it for sure cuts down on copies compared to before.

jonasschnelli commented:
Concept ACK.

laanwj commented:
winsok needs an extra cast:
```
../../src/net.cpp: In function ‘size_t SocketSendData(CNode*)’:
../../src/net.cpp:777:131: error: invalid conversion from ‘const value_type* {aka const unsigned char*}’ to ‘const char*’ [-fpermissive]
         int nBytes = send(pnode->hSocket, &data[pnode->nSendOffset], data.size() - pnode->nSendOffset, MSG_NOSIGNAL | MSG_DONTWAIT);    
```

theuni committed:
net: Set feelers to disconnect at the end of the version message

This way we're not relying on messages going out after fDisconnect has been
set.

This should not cause any real behavioral changes, though feelers should
arguably disconnect earlier in the process. That can be addressed in a later
functional change.

theuni commented:
Fixed several things:
- Fixed @jonasschnelli's whitespace nit
- Fixed win32 build, broken due to missing cast
- Addressed @TheBlueMatt's complaints:
 - Back to using ```pnode->nVersion == 0``` to detect lack of version message
 - Added a change to hold off on setting fDisconnect for feelers until later, to avoid attempting to send messages to a disconnected node. The behavior is kept the same here, but imo it should be changed as a follow-up.
  - zero copies when sending to the net. CSerializedNetMsg now advises how much padding it may possibly need. Its copy ctor and copy assignment operator are deleted to ensure no copies.
   - CVectorInserter::pad and CVectorOverWriter::skip. I'm not sure if these are exactly what you needed, but I took a guess :)

laanwj commented:
I'm testing this + #9125

sipa commented:
Needs rebase after #9075.

sipa commented on src/streams.h:
```diff
@@ -69,6 +69,98 @@ OverrideStream<S> WithOrVersion(S* s, int nVersionFlag)
+    {
+        return nType;
+    }
+    void skip(size_t nSize)
```
CSizeComputer uses 'seek' for this function (inspired by the dd command, where skip is something you do on input, and seek is something you do on output).

theuni replied:
Makes sense. Will change.

theuni commented on src/streams.h:
```diff
@@ -69,6 +69,98 @@ OverrideStream<S> WithOrVersion(S* s, int nVersionFlag)
+ *
+ * Note that the referenced vector must already be large enough to hold all data to be serialized.
+ */
+class CVectorOverWriter
```
Yes, I think that should work fine. Will change.

TheBlueMatt commented on src/net.cpp:
```diff
@@ -2581,30 +2581,17 @@ void CNode::AskFor(const CInv& inv)
-
-    unsigned int nSize = strm.size() - CMessageHeader::HEADER_SIZE;
-    LogPrint("net", "sending %s (%d bytes) peer=%d\n",  SanitizeString(sCommand.c_str()), nSize, pnode->id);
+    uint256 hash = Hash(msg.data.data() + msg.header_pad_size, msg.data.data() + msg.data.size());
```
Shouldnt this happen inside CNetMsgMaker::Make? That way we can just have a BIP151MsgMaker?

theuni replied:
It was left out of CNetMsgMaker because I assumed that we'd want CNode making the decision about how to wrap the data for sending. If it's to be encrypted, we'll need some per-node state vars.

So I guess it depends whether encryption status and state are to end up somewhere in main like CNodeState, or in CNode. Since (for once) this seems like something CConnman should know about but nobody else should have to care, I figured CNode was the likely place.

theuni replied:
I should mention: I've gone back and forth a few times on this (probably in this PR even). If anyone has a strong preference for a different approach that makes sense, I'm fine with changing it.

TheBlueMatt replied:
This isnt in CNode, though? If it were in CNode I'd be OK with that, but putting it in CConnman seems like /the/ layer violation this PR intends to fix?

theuni replied:
The result of this PR is that CConnman now receives only dumb bytes with no knowledge of what they are. All internal structures could change without affecting the net side.

Dealing with wrapping the raw bytes up for transmission is something different, and arguably not something that the bitcoin side should have to know/care about. If you have a different model in your mind, please suggest it.

CNode is slowly moving towards an internal dumb storage class. So in this case, I don't make much distinction.

TheBlueMatt commented on src/net.cpp:
```diff
@@ -2581,30 +2581,17 @@ void CNode::AskFor(const CInv& inv)
 }
 
-CDataStream CConnman::BeginMessage(CNode* pnode, int nVersion, int flags, const std::string& sCommand)
+void CConnman::PushMessage(CNode* pnode, CSerializedNetMsg msg)
```
Shouldn't this be CSerializedNetMsg&&? It seems super strange that we're secretly destroying the object passed in, but only because the CSerializedNetMsg class doesnt have a copy-constructor.

sipa replied:
Nothing secret is being done. This would fail, for example:
```c++
CSerializedNetMsg x = ...;
pconman->PushMessage(pnode, x);
```
as the binding of the actual parameter x would need a copy constructor to assign to the formal parameter msg.

theuni replied:
It was done this way originally so that a message could be serialized once and sent to multiple peers. If an rvalue is required, we can't do that.

I don't think there's a use-case for that at the moment, though, so changing to && is fine by me.

TheBlueMatt commented on src/netmessagemaker.h:
```diff
@@ -0,0 +1,37 @@
+    {
+        CSerializedNetMsg msg;
+        msg.data.assign(CSerializedNetMsg::header_pad_size, 0);
+        msg.command = std::move(sCommand);
```
Over-use of std::move, much? Can we stick with const std::string& in the function declaration and then use operator=() to make the copy instead of creating a new object just to std::move the data over?
(Also, could definitely be missing some part of rvalue stuff, here)

theuni replied:
This is standard c++11 pass-by-value-by-default. If you pass in a const-reference, a copy is guaranteed since we're storing the result. If you pass by value, it can be moved and the copy can be skipped.

Not sure why you'd prefer a copy over 2 cheap moves.

Arguably NetMsgType should become a
```std::array<char, CMessageHeader::COMMAND_SIZE>```
so we can do away with the char arrays/strings, but that's a change for another day.

TheBlueMatt reviewed (commented):
Again didnt do a full review, just looked over the actual code changes and came up with a few comments...will do a full review when they're addressed.

theuni commented:
Rebased and updated for comments from @sipa and @TheBlueMatt.
- Combined the vector serializers into CVectorWriter as requested.
- skip() -> seek()
- Added tests for CVectorWriter.
- PushMessage requires an rvalue

sipa commented:
Hmm, I would consider CNode and CConnman to be the same layer (after all
its processing related stated is moved out, and the rest is made private to
the net module).

TheBlueMatt commented:
Heh, well if you consider CNode to be a part of CConnman, then the type-of-message-to-send logic should be in net_processing...either way CConnman shouldnt know anything about the serialization of messages on the wire.

MarcoFalke commented:
Needs rebase

laanwj commented:
Code review ACK 02d463d
Have been testing an earlier revision on a node for 6 days without issues.

theuni commented:
updated after a discussion with @TheBlueMatt and @sipa today, I think we're all finally aligned.

@laanwj Sadly removing the "skb" was part of the compromise as it was a slight layer violation. We instead use 2 send()s (later scatter/gather) to avoid the copy.

sipa commented:
Needs rebase, sorry!

theuni commented:
Rebased after ca8549d2bd32f17f8b69d1edbe3f2976fba504b4.

sipa commented:
utACK 09a92c27d46a62d10aed95b742e891e204dcee4e

paveljanik commented:
ACK 09a92c2

TheBlueMatt commented on src/streams.h:
```diff
@@ -69,6 +69,75 @@ OverrideStream<S> WithOrVersion(S* s, int nVersionFlag)
+ * @param[in]  nVersionIn Serialization Version (including any flags)
+ * @param[in]  vchDataIn  Referenced byte vector to overwrite/append
+ * @param[in]  nPosIn Starting position. Vector index where writes should start. If the index is past the end, the
+ *                    vector will first be resized to fit. So to append, use vec.size().
```
This comment is very confusing...if nPos is 1 and vchData.size() is 1, the documentation implies that vchData will be resized so that nPos isnt past-the-end, but it will not be. Instead, vchData is resized so that, at maximum, nPos is 1-past-the-end.

theuni replied:
Hmm, yes. Will clarify.

TheBlueMatt commented on src/net.cpp:
```diff
@@ -2612,30 +2612,19 @@ void CNode::AskFor(const CInv& inv)
-}
+    size_t nMessageSize = msg.data.size();
+    size_t nTotalSize = nMessageSize + CMessageHeader::HEADER_SIZE;
+    LogPrint("net", "sending %s (%d bytes) peer=%d\n",  SanitizeString(msg.command.c_str()), nTotalSize, pnode->id);
```
You changed the print - should print nMessageSize, not nTotalSize.

theuni replied:
Good call. I assumed it'd never say "sending 0 bytes", but you're right, I'll avoid a behavioral change here.

TheBlueMatt reviewed (commented):
You missed one fDisconnect check in SendMessages (the last one - feefilter, maybe a rebase issue?).

theuni commented:
Updated and squashed for @TheBlueMatt's comments. The diff from before is very straightforward:
```diff
cory@cory-i7:~/dev/bitcoin/src(connman-send)$ git diff theuni/connman-send
diff --git a/src/main.cpp b/src/main.cpp
index 4520afb..5595381 100644
--- a/src/main.cpp
+++ b/src/main.cpp
@@ -6983,7 +6983,7 @@ bool SendMessages(CNode* pto, CConnman& connman)
         // Message: feefilter
         //
         // We don't want white listed peers to filter txs to us if we have -whitelistforcerelay
-        if (!pto->fDisconnect && pto->nVersion >= FEEFILTER_VERSION && GetBoolArg("-feefilter", DEFAULT_FEEFILTER) &&
+        if (pto->nVersion >= FEEFILTER_VERSION && GetBoolArg("-feefilter", DEFAULT_FEEFILTER) &&
             !(pto->fWhitelisted && GetBoolArg("-whitelistforcerelay", DEFAULT_WHITELISTFORCERELAY))) {
             CAmount currentFilter = mempool.GetMinFee(GetArg("-maxmempool", DEFAULT_MAX_MEMPOOL_SIZE) * 1000000).GetFeePerK();
             int64_t timeNow = GetTimeMicros();
diff --git a/src/net.cpp b/src/net.cpp
index e05c174..ce87150 100644
--- a/src/net.cpp
+++ b/src/net.cpp
@@ -2616,7 +2616,7 @@ void CConnman::PushMessage(CNode* pnode, CSerializedNetMsg&& msg)
 {
     size_t nMessageSize = msg.data.size();
     size_t nTotalSize = nMessageSize + CMessageHeader::HEADER_SIZE;
-    LogPrint("net", "sending %s (%d bytes) peer=%d\n",  SanitizeString(msg.command.c_str()), nTotalSize, pnode->id);
+    LogPrint("net", "sending %s (%d bytes) peer=%d\n",  SanitizeString(msg.command.c_str()), nMessageSize, pnode->id);
 
     std::vector<unsigned char> serializedHeader;
     serializedHeader.reserve(CMessageHeader::HEADER_SIZE);
diff --git a/src/streams.h b/src/streams.h
index ef22620..b508784 100644
--- a/src/streams.h
+++ b/src/streams.h
@@ -81,8 +81,8 @@ class CVectorWriter
  * @param[in]  nTypeIn Serialization Type
  * @param[in]  nVersionIn Serialization Version (including any flags)
  * @param[in]  vchDataIn  Referenced byte vector to overwrite/append
- * @param[in]  nPosIn Starting position. Vector index where writes should start. If the index is past the end, the
- *                    vector will first be resized to fit. So to append, use vec.size().
+ * @param[in]  nPosIn Starting position. Vector index where writes should start. The vector will initially
+ *                    grow as necessary to  max(index, vec.size()). So to append, use vec.size().
 */
     CVectorWriter(int nTypeIn, int nVersionIn, std::vector<unsigned char>& vchDataIn, size_t nPosIn) : nType(nTypeIn), nVersion(nVersionIn), vchData(vchDataIn), nPos(nPosIn)
     {
```

TheBlueMatt commented:
ACK c7be56dceff625991cb2884f2ce053996ac613cd
